### ISSUE TITLE
Assertion failure in nlpMDS_ex4.exe

### ISSUE THREAD

**Issue by CameronRutherford** [LLNL/hiop#134]
On branch [`develop`](https://github.com/LLNL/hiop/tree/develop), running `mpirun -n 1 nlpMDS_ex4.exe 400 100 1 -selfcheck` results in an assertion failure:
```shell
nlpMDS_ex4.exe: /<redacted>/hiop-git/src/Optimization/hiopPDPerturbation.hpp:184: \
bool hiop::hiopPDPerturbation::compute_perturb_singularity(double&, double&, double&, double&): \
Assertion `delta_cc == 0. && delta_cd == 0.' failed.
```
This issue should be resolved before #103 can be resolved.

CC: @pelesh @ashermancinelli

**Comment by ashermancinelli** [comment]
What build configuration did you use to find this bug?

**Comment by cnpetra** [comment]
> What build configuration did you use to find this bug?

Was `HIOP_USE_GPU` on? I cannot reproduce. 

btw, looking at #103 is at the top of my priorities...

**Comment by CameronRutherford** [comment]
I did not change any of the default configurations, as I was building using the build script on Newell. What I did notice is that when running on a partition with a GPU requested, I was not able to re-produce the error, while running on a partition with a GPU resulted in no error.

i.e., without a GPU (on Newell01) the following is output:

```shell
$ mpirun -n 1 ./nlpMDS_ex4.exe 400 100 1 -selfcheck
[NVBLAS] No Gpu available
[NVBLAS] NVBLAS_CONFIG_FILE environment variable is set to '/qfs/projects/exasgd/newell/nvblas.conf'
...
[Warning] There is no reason to set 'acceptable_tolerance' tighter than 'tolerance'. Will set the two to 'tolerance'.
...
[Error] hiopLinSolverMagmaBuka error: 113 argument to dsytrf has an illegal value.
[Error] hiopLinSolverMagmaBuka error: 113 argument to dsytrf has an illegal value.
nlpMDS_ex4.exe: /<redacted>/hiop-git/src/Optimization/hiopPDPerturbation.hpp:184: bool hiop::hiopPDPerturbation::compute_perturb_singularity(double&, double&, double&, double&): Assertion `delta_cc == 0. && delta_cd == 0.' failed.
[newell01:28193] *** Process received signal ***
```

Running on Newell03 with a GPU has this passing. Perhaps this is an anomaly...

**Comment by ashermancinelli** [comment]
> ... while running on a partition without a GPU resulted in an error.

Are you saying building with `HIOP_USE_GPU=ON` and running without a GPU available fails? If so, I think this is common sense and not really a bug.

**Comment by CameronRutherford** [comment]
I suppose so. It seems odd to me that running code that should be executed only on the CPU requires a GPU to be available, but it is also completely reasonable to require people to build with `HIOP_USE_GPU=OFF` in the first place.

**Comment by ashermancinelli** [comment]
> It seems odd to me that running code that should be executed only on the CPU requires a GPU to be available

If you run the driver after building with GPU=ON, it will use MAGMA if not also RAJA/Umpire. Are you sure this shouldn't run on the device?

**Comment by CameronRutherford** [comment]
I think I misunderstood the nature of the driver `nlpMDS_ex4.exe`. 

Initially, I built with `HIOP_USE_GPU=ON`, and then attempted to run said driver without a GPU being available. Since this Driver does use the GPU when `HIOP_USE_GPU=ON` is set, I was getting the errors that I posted in the initial comment.

I was under the impression that this driver does not require the GPU to run regardless of the state of `HIOP_USE_GPU`, however it seems as though I was incorrect with this assumption. So it makes sense that I would require a GPU to be available in order to run the specified driver given the build options that I was using.

Does this help clarify?
